File: java/app/src/main/java/apache_flink/kickstarter/KafkaClientPropertiesLookup.java
/**
 * Copyright (c) 2024 Jeffrey Jonathan Jennings
 * 
 * @author Jeffrey Jonathan Jennings (J3)
 * 
 * 
 * An Apache Flink custom source data stream is a user-defined source of data that
 * is integrated into a Flink application to read and process data from non-standard
 * or custom sources. This custom source can be anything that isn't supported by Flink
 * out of the box, such as proprietary REST APIs, specialized databases, custom hardware 
 * interfaces, etc. J3 utilizes a Custom Source Data Stream to read the AWS Secrets Manager 
 * secrets and AWS Systems Manager Parameter Store properties during the initial start of a 
 * App, then caches the properties for use by any subsequent events that need these properties.
 */
package apache_flink.kickstarter;

import org.apache.flink.api.common.functions.RichMapFunction;
import org.apache.flink.configuration.Configuration;
import java.util.concurrent.atomic.AtomicReference;
import java.util.*;

import apache_flink.kickstarter.helper.*;


/**
 * This class creates a Custom Source Data Stream to read the AWS Secrets Manager secrets 
 * and AWS Systems Manager Parameter Store properties during the initial set up of the 
 * Flink App, then caches the properties for use by any subsequent events that need these
 * properties.
 */
public class KafkaClientPropertiesLookup extends RichMapFunction<Properties, Properties>{
    private transient AtomicReference<Properties> _properties;
    private volatile boolean _consumerKafkaClient;
    private volatile String _serviceAccountUser;


    /**
     * Default constructor.
     * 
     * @param consumerKafkaClient
     * @param serviceAccountUser
     * @throws Exception - Exception occurs when the service account user is empty.
     */
    public KafkaClientPropertiesLookup(final boolean consumerKafkaClient, final String serviceAccountUser) throws Exception {
        // --- Check if the service account user is empty
        if(serviceAccountUser.isEmpty()) {
            throw new Exception("The service account user must be provided.");
        }

        // ---  Set the class properties
        this._consumerKafkaClient = consumerKafkaClient;
        this._serviceAccountUser = serviceAccountUser;
    }

    /**
     * This method is called once per parallel task instance when the job starts. 
     * 
     * @parameters The configuration containing the parameters attached to the
     * contract.
     * @throws Exception - Implementations may forward exceptions, which are caught
     * by the runtime.  When the runtime catches an exception, it aborts the task and 
     * lets the fail-over logic decide whether to retry the task execution.
     */
    @Override
    public void open(Configuration configuration) throws Exception {
        /* 
         * Get the Kafka Client properties from AWS Secrets Manager and AWS Systems
         * Manager Parameter Store.
         */
        final String secretPathPrefix = "/confluent_cloud_resource/" + this._serviceAccountUser;
        final KafkaClient kafkaClient = 
            new KafkaClient(
                secretPathPrefix + "/kafka_cluster/java_client", 
                secretPathPrefix + (this._consumerKafkaClient ? "/consumer_kafka_client" : "/producer_kafka_client"));
        ObjectResult<Properties> properties = kafkaClient.getKafkaClusterPropertiesFromAws();

		if(!properties.isSuccessful()) { 
			throw new RuntimeException("Failed to retrieve the Kafka Client properties could not be retrieved because " + properties.getErrorMessageCode() + " " + properties.getErrorMessage());
		} else {
            // ---  Set the class properties
            this._properties = new AtomicReference<>(properties.get());
        }
    }

    /**
     * This method is called for each element of the input stream.
     * 
     * @param value - The input value.
     * @return The result of the map operation.
     */
    @Override
    public Properties map(Properties value) {
        return(this._properties.get());
    }
    
    /**
     * This method is called when the task is canceled or the job is stopped.
     * 
     * @throws Exception - Implementations may forward exceptions, which are
     * caught.
     */
    @Override
    public void close() throws Exception {}
}

File: python/kickstarter/kafka_client_properties_lookup.py
from pyflink.common import Configuration
from pyflink.datastream import SourceFunction
from threading import Lock

from helper.kafka_client import KafkaClient

__copyright__  = "Copyright (c) 2024 Jeffrey Jonathan Jennings"
__credits__    = ["Jeffrey Jonathan Jennings"]
__license__    = "MIT"
__maintainer__ = "Jeffrey Jonathan Jennings"
__email__      = "j3@thej3.com"
__status__     = "dev"


class KafkaClientPropertiesLookup(SourceFunction):
    """
    An Apache Flink custom data source stream is a user-defined source of data that
    is integrated into a Flink application to read and process data from non-standard
    or custom sources. This custom source can be anything that isn't supported by Flink
    out of the box, such as proprietary REST APIs, specialized databases, custom hardware
    interfaces, etc. This code uses a Custom Data Source Stream to read the AWS Secrets 
    Manager secrets and AWS Systems Manager Parameter Store properties during the initial
    start of the Flink App, then caches the properties for use by any subsequent events
    that need these properties.

    Args:
        SourceFunction (obj): In Apache Flink, the SourceFunction class is a rich variant of the
        SourceFunction class. It provides access to the RuntimeContext and includes setup and
        teardown methods. 
    """

    def __init__(self, flag, options):
        self.flag = flag
        self.options = options
        
        if not options.get('s3_bucket_name'):
            raise Exception("The 's3_bucket_name' must be provided.")

        # Set the class properties
        self._is_consumer = options.get('is_consumer')
        self._service_account_user = options.get('s3_bucket_name')

        self._properties = {}  # This acts like Java's Properties class

        self.is_running = True

    def run(self, ctx):
        """
        This method is called once per parallel task instance when the job starts.
        It gets the Kafka Client properties from AWS Secrets Manager and
        AWS Systems Manager Parameter Store, then stores the properties in class attributes.        
        """

        # Get the Kafka Client properties from AWS Secrets Manager and AWS Systems Manager Parameter Store.
        secret_path_prefix = f"/confluent_cloud_resource/{self._service_account_user}"
        kafka_client = KafkaClient(
            f"{secret_path_prefix}/kafka_cluster/java_client",
            f"{secret_path_prefix}/consumer_kafka_client" if self._is_consumer else f"{secret_path_prefix}/producer_kafka_client"
        )
        properties = kafka_client.get_kafka_cluster_properties_from_aws()

        if not properties.is_successful():
            raise RuntimeError(f"Failed to retrieve the Kafka Client properties from '{secret_path_prefix}' secrets because {properties.get_error_message_code()}:{properties.get_error_message()}")
        else:
            ctx.collect(properties.get())
            self.is_running = False
    
    def cancel(self):
        self.is_running = False